### PR TITLE
ENH: Geometric quality metrics for random sampling 

### LINKED ISSUE #14037: ENH: other quality metrics for random sampling
The new `scipy.stats.qmc` sub-module provide `discrepancy` as a way to measure the quality of a random sample. This measure is a uniformity criterion.

Other metrics can be used to measure the quality such as _mindist_ and _minimum spanning tree_. 

Here is a simple implementation of useful measures:

```python
import numpy as np
from scipy.spatial.distance import cdist


def l1_norm(sample):
    l1 = cdist(sample, sample, 'cityblock')
    return np.min(l1[l1.nonzero()])


# mindist
def l2_norm(sample):
    l2 = cdist(sample, sample)
    return np.min(l2[l2.nonzero()])


def linf_norm(sample):
    linf = cdist(sample, sample, 'chebyshev')
    return np.min(linf[linf.nonzero()])
```

These are more "geometrically/topologically" based whereas the discrepancy is a statistical criterion. We could have a common function with a `method` argument to select one of the metrics.

From a dependency perspective, this should go in `scipy.spatial`. Also, if we add _mst_, `scipy.spatial` already has a dependency to `scipy.sparse`.

### PR DESCRIPTION
**Pull request by aglebov**
#### Reference issue
Closes #14037

#### What does this implement/fix?
The change adds simple geometric/topological metrics for random sampling.

#### Additional information
The metrics are implemented in a new function rather than added to the existing `scipy.stats.qmc.discrepancy` function as they represent a different type of metrics: distance/geometric-based vs distribution-based. The interpretation of the returned value is also different between the two functions. See the discussion in the comments below (particularly, https://github.com/scipy/scipy/pull/18893#pullrequestreview-1531761774 and https://github.com/scipy/scipy/pull/18893#issuecomment-1650105609) for further details.

### CODE REVIEW

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:323]:
```suggestion
    metric: str or callable, optional
```

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:330]: This could use a small Notes section explaining a bit the metrics, an Example section with 1-2 examples (bonus point if you make examples with the tree for mst and packing circles for mindist. I have a gist you can use: https://gist.github.com/tupui/2719245158f2dc2366a9256b58a85519). And lastly can you add a reference section with 1-2 articles?

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:341]: We can also add an input validation for `sample`. We could allow to pass something else than a unit hypercube though. Also to question as we could want to have consistency with the other metric.

**Review by tupui — CHANGES_REQUESTED** [review]
This is a great start @aglebov 🚀 

To your questions:

> We had a few questions:
> 
> Should we add this function in a new module in scipy.spatial, keep it in scipy.stats or perhaps merge it into scipy.stats.qmc.discrepancy?

I think `scipy.stats.qmc` is the right place, especially if we keep this name (I think it's a good one.)

I would not mix these method into `scipy.stats.qmc.discrepancy` as it's different type of metrics: distribution based vs distance/geometric based.

> We can get two metrics from the minimum spanning tree: the mean distance and the standard deviation, so we could let the user choose which metric they want or return both.

I would just return the mean for now. (I am considering the median, but it might actually be good to get the bias from outliers to worsen the metric. Open to discussion I would say.)

Only returning a single value would keep the return value consistent with `scipy.stats.qmc.discrepancy`.

Besides that, it will mostly need tests. Typically we try to find something from the literature, but I understand that here it might be tricky. It would be ok to have a few canonical cases (like some points in corners or at very specific coordinates so it's easy to understand), and then we can have a test which is checking that the metrics improve when the quality of a sample improves. You can have a look at the current tests for `scipy.stats.qmc.discrepancy`.

**Comment by mdhaber** [comment]
@aglebov looks like you have a review waiting, so I thought I'd check in. Do you plan to finish this?

**Comment by aglebov** [comment]
We also wanted to get a view on the issue with minimum spanning tree (#18892) before moving forward with this change. Currently, `scipy.sparse.csgraph.minimum_spanning_tree` is unable to distinguish between the absence of an edge and an edge with zero weight in the input. As a consequence, if we apply our current implementation to a sample containing duplicate points, `minimum_spanning_tree` will ignore the edges with 0 distances and overestimate the quality of the sample.

**Comment by tupui** [comment]
I would be fine with leaving this case out for now.

I am thinking though, having duplicate points could be seen as very bad and we could raise a warning in general.

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done, I've added the sample validations as used in `discrepancy`, including the check that the sample is within a unit hypercube.

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py:384]: Added notes, references and examples.

> bonus point if you make examples with the tree for mst and packing circles for mindist.

Do you mean examples with specific arrangements of points?

**Comment by tupui** [comment]
So I was thinking about that and maybe we could consider normalising the value with the size of the domain? This way we can have a value between 0 and 1 and make it so that the interpretation is the same as `discrepancy`.

Another possibility would be to add a negative sign, but that sounds less elegant.

What do you think?

---

Unrelated, I am curious. Are you a current user of this submodule? Or a QMC practitioner?

**Comment by aglebov** [comment]
> I am thinking though, having duplicate points could be seen as very bad and we could raise a warning in general.

I've added a warning now.

> So I was thinking about that and maybe we could consider normalising the value with the size of the domain? This way we can have a value between 0 and 1 and make it so that the interpretation is the same as discrepancy.

If we divide by some positive number, then higher values would still correspond to more uniform samples (unlike in `discrepancy`), unless I'm not following the direction of your thought here.

The return value of `discrepancy` does not seem to be bounded either:
```
>>> s = np.random.uniform(size=(100, 100))
>>> qmc.discrepancy(s)
55451062.391446136
```

> Unrelated, I am curious. Are you a current user of this submodule? Or a QMC practitioner?

It's the first time I've used this submodule - we picked #14037 during the sprint at SciPy 2023.

**Comment by tupui** [comment]
Thank you for the updates @aglebov 🚀 

I will reply to your points tomorrow when I have more time 😃

For future update, please add new commits instead (and if you need or want to update your branch with respect to `main` only use merge and not rebase.) This is because GitHub is not happy with history being overwritten. i.e. the emails that GitHub send have links which gets outdated. e.g.

Screenshot 2023-07-25 at 20 12 10

**Comment by tupui** [comment]
> If we divide by some positive number, then higher values would still correspond to more uniform samples (unlike in `discrepancy`), unless I'm not following the direction of your thought here.

Since the output is a distance in the hypercube, it can be divided by the maximum distance in the hypercube (which depends on the distance metric, 1 for Euclidean distance). Then we would do `1-d` to reverse the ordering.

> The return value of `discrepancy` does not seem to be bounded either:

Yeah this is a strange "animal". We could find some meaning for some discrepancy but I would not want to write anything in the doc because I would not be sure if that would help the comprehension 😅 There are also, in some cases, upper/lower bounds for the measures, but I never really took the time to see if I could implement these.

> It's the first time I've used this submodule - we picked #14037 during the sprint at SciPy 2023.

Lucky me then, thank you! 😃 You are making great comments and progress so I assumed you were somehow familiar with the topic.

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:333]: We can do the same in `discrepancy` to link to this function. (GitHub does not allow me to make a suggestion "that far".)
```suggestion
    Returns
    -------
    discrepancy : float
        Discrepancy (higher values correspond to greater sample uniformity).

	See Also
    --------
    discrepancy

    Notes
    -----
```

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:346]:
```suggestion
    with higher mean and lower standard deviation being preferable (see [1]_
```

File: scipy/stats/_qmc.py
Comment (tupui) [reply scipy/stats/_qmc.py:383]: Thank you!

For the figures, I was thinking about things like that:

This one is the output of the gist I shared
43350899-9efe0f08-920b-11e8-8a20-e9dab98f7a52

And this is a maximin on the left and a minimax sampling on the right. Here I am just saying we could have a circle around the "critical point" to show the minimal distance (the circle would touch the nearest point then).
Figure-Maximin-left-and-Minimax-right-designs_W640

Comment (tupui) [reply scipy/stats/_qmc.py:383]: Nice 👍

I was actually thinking about 2 different figures, one per method. But with a legend, having both on the same figure could work and save some space.

I would set the aspect ratio to 1 so that the circle is really a circle.

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:372]: Since we are in the QMC module, I would prefer we use a QMC sample, `LatinHypercube` for instance to have a bit of variation.

For the RNG we also need to use the `Generator` interface from NumPy.

Then the values bellow will have to be updated. If you need any help, I am happy to pull your branch and give a hand!

```suggestion
    >>> rng = np.random.default_rng(191468432622931918890291693003068437394)
    >>> sample = qmc.LatinHypercube(d=2).random(50)
```

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:396]:
```suggestion
    >>> _ = ax.set(aspect='equal', xlabel=r'$x_1$', ylabel=r'$x_2$',
    ...            xlim=[0, 1], ylim=[0, 1])
```

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:388]:
```suggestion
    >>> from scipy.spatial.distance import pdist, squareform
```

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:394]:
```suggestion
    >>> dist = squareform(pdist(sample))
    >>> mst = minimum_spanning_tree(dist)
    >>> edges = np.where(mst.toarray() > 0)
    >>> edges = np.asarray(edges).T
    >>> min_dist = np.min(dist)
    >>> min_idx = np.argwhere(dist == min_dist)[0]
```

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:405]: Trying to make it fit
```suggestion
    ...     Line2D([0], [0], marker='o', lw=0, label='Sample'),
    ...     Line2D([0], [0], color='k', label='MST'),
    ...     Line2D([0], [0], marker='o', lw=0, markerfacecolor='w', markeredgecolor='r',
    ...            label='mindist'),
```

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:418]: More a note to myself for a follow up (Or you if you are interested of course 😄). We are starting to duplicate this. We should make a helper function with that snippet.

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:326]:
```suggestion
    method : {"mindist", "mst"}, optional
        The method to use. One of ``mindist`` for minimum distance
        or ``mst`` for minimum spanning tree.
```

File: scipy/stats/_qmc.py
Comment (tupui) [inline scipy/stats/_qmc.py:329]: Single ticks should resolve the link
```suggestion
        for `scipy.spatial.distance.pdist`.
```

**Review by tupui — CHANGES_REQUESTED** [review]
Thank you for the updates @aglebov. This is in good shape and besides the open discussion around the metric being high or low, this will just be missing some tests.

For the name, `geometric_uniformity` is also fine though I think I prefer `geometric_discrepancy` just because there is more unity in the names then. I would accept both.

Let me know if you need help with anything. FYI after next week I will be away for a month (without internet probably too.)

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done, though I find the resulting legend a bit cryptic

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Unfortunately, if we do this:
```
dist = squareform(pdist(sample))
```
then `np.min` will always give us 0 due to diagonal elements:
```
min_dist = np.min(dist)
```

I've changed the rest.

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py]: Done

**Comment by aglebov** [comment]
> Since the output is a distance in the hypercube, it can be divided by the maximum distance in the hypercube (which depends on the distance metric, 1 for Euclidean distance). Then we would do `1-d` to reverse the ordering.

It would be ideal to get a view from somebody who has experience using these metrics. To me, the "raw" mean or minimum distance seem easier to interpret than if they were scaled and then subtracted from 1.

**Comment by tupui** [comment]
> It would be ideal to get a view from somebody who has experience using these metrics. To me, the "raw" mean or minimum distance seem easier to interpret than if they were scaled and then subtracted from 1.

That would be me 😅 But yes, I will reach out to folks to get of sense of what they think is best. I don't think there is a lot of prior use of these. This part (same goes with the discrepancy) is fairly obscure and used to be limited to academics for not so long ago.

File: scipy/stats/_qmc.py
Comment (tupui) [reply scipy/stats/_qmc.py]: I am happy to do otherwise if you find a way for the legend to not being cut.

Not sure if you've seen that you can check here the rendered documentation. You just have to click on the job which reads `Check rendered doc...` once it's green/passing.

Also taking the opportunity here to ask you to put these in your commit messages until we have tests: `[skip actions] [skip cirrus]`. This will just let CircleCI run to build the doc. And once we just want the tests to run, you will be able to do `[skip circle]`. Thanks!!

File: scipy/stats/_qmc.py
Comment (tupui) [reply scipy/stats/_qmc.py]: Ah yes 👍

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py:430]: That should be done.

My problem was that running `make html` does not seem to regenerate the plots resulting from docstrings. Is there some trick to do it (other than deleting `doc/build` and generating the whole thing from scratch, which takes ages...)?

Comment (tupui) [reply scipy/stats/_qmc.py:430]: That's strange, but Sphinx... What you can try is to delete the output html page and Sphinx should detect that it has to rebuild it.

Comment (aglebov) [reply scipy/stats/_qmc.py:430]: Unfortunately, deleting the generated `html` or `png` files seems to have no effect - the image seems to be cached somewhere else.

Comment (tupui) [reply scipy/stats/_qmc.py:430]: mmm ok, and by `png` you tried to remove `doc/build/plot_directive/reference/generated/fig.png` or just what was in the `html` folder?

Comment (aglebov) [reply scipy/stats/_qmc.py:430]: I tried to remove the figure in the `html` directory. Removing it from the `plot_directive` does the trick, thanks!

Once I did this another interesting bit of behaviour was revealed. The code of the docstring for `geometric_discrepancy` contains this line:
```
>>> rng = np.random.default_rng(191468432622931918890291693003068437394)
```
whereas the rendered html becomes:
```
>>> rng = np.random.default_rng()
```
and a different graph is created each time.

I was not aware that something would try to erase the seed:)

Comment (tupui) [reply scipy/stats/_qmc.py:430]: Yes this is some magic (I actually introduced) to not present a seed in the doc. We don't want users to copy paste the seed value and in the end transform a RNG into a fixed/bad "NG".

Before we used to have a mechanism that would inject the seed into `>>> rng = np.random.default_rng()` but this is really not convenient for developers as you need to first put the seed on your side, run the doctests, get the right numbers and then remove the seed. And also if you would run the doctests manually on your side with an IDE, that would fail.

With the new mechanism, you can in the code put the seed you want. And still, end users reading the doc would not see any seed.

There is a bit more details there https://scipy.github.io/devdocs/dev/contributor/rendering_documentation.html#examples-and-randomness

Comment (tupui) [reply scipy/stats/_qmc.py:430]: Pleasure 😃

To be clear, the seed is removed from the rendered HTML, but it is still used when building the figure.

Note that I would not mind having a varying figure in that case. Also here an alternative would be to hand pick a design of something like 10-20 points if we really want to tell a story about the points themselves, etc.

Comment (aglebov) [reply scipy/stats/_qmc.py:430]: > To be clear, the seed is removed from the rendered HTML, but it is still used when building the figure.

That's not what I see though. Every time I regenerate it, I get a different figure.

Not a big deal of course, just a bit surprising...

**Comment by aglebov** [comment]
> Besides that, it will mostly need tests. Typically we try to find something from the literature, but I understand that here it might be tricky. It would be ok to have a few canonical cases (like some points in corners or at very specific coordinates so it's easy to understand), and then we can have a test which is checking that the metrics improve when the quality of a sample improves. You can have a look at the current tests for scipy.stats.qmc.discrepancy.

I've added some simple tests for `geometric_discrepancy` in the latest commit.

If you have any ideas for a good example to demonstrate the improvement in the quality, please let me know.

One thing that came to mind on that point is that `geometric_discrepancy` can only be used for comparing the quality of samples of the same size: for samples of equal size, larger minimum distance between points is better, whereas  increasing the sample size within a unit hypercube makes the minimum distance smaller.

File: scipy/stats/tests/test_qmc.py
Comment (tupui) [inline scipy/stats/tests/test_qmc.py:195]: This should fix the CI
```suggestion
        with pytest.warns(UserWarning, match="Sample contains duplicate points."):
```

File: scipy/stats/tests/test_qmc.py
Comment (tupui) [inline scipy/stats/tests/test_qmc.py:224]: Yes so this part could instead look like what is done in `test_lloyd`. 4 iterations are done with an increasing number of samples and the metrics should improve.

File: scipy/stats/tests/test_qmc.py
Comment (tupui) [inline scipy/stats/tests/test_qmc.py:202]: These look good 👍 Could be parametrised but this is also fine as it is as there are not that many cases.

**Review by tupui — COMMENTED** [review]
Thank you for adding the tests! This is great, I just have some minor comments.

File: scipy/stats/tests/test_qmc.py
Comment (tupui) [reply scipy/stats/tests/test_qmc.py:224]: I am wrong, discard that comment.

**Comment by tupui** [comment]
> One thing that came to mind on that point is that geometric_discrepancy can only be used for comparing the quality of samples of the same size: for samples of equal size, larger minimum distance between points is better, whereas increasing the sample size within a unit hypercube makes the minimum distance smaller. I've added a note to the docstring on this point.

Right 👍 that's a good note to add in the doc

> If you have any ideas for a good example to demonstrate the improvement in the quality, please let me know.

I thought I had something but actually no 😅 Besides comparing two samples where we know which one is better, I am not sure. Also I don't think it would be really safe to use the other discrepancies to see if these new measures follow the same trend. They don't work the same way and that would not be a really useful test to me.

I think we are good like that. You tested a few canonical cases, and the code is simple enough so that I am not sure anything else could happen that would justify another test. If you have other ideas though please feel free to share your thoughts.

File: scipy/stats/_qmc.py
Comment (aglebov) [reply scipy/stats/_qmc.py:319]: > Sorry if you understood I was requesting the change for your PR to move forward.

You were pretty clear it was a note to self:)

File: scipy/stats/tests/test_qmc.py
Comment (aglebov) [reply scipy/stats/tests/test_qmc.py]: Done

File: scipy/stats/tests/test_qmc.py
Comment (aglebov) [reply scipy/stats/tests/test_qmc.py]: Done

**Comment by aglebov** [comment]
> Right 👍 that's a good note to add in the doc

That should be there already:

```
    Also note that when comparing different sampling strategies using this function,
    the sample size must be kept constant.
```

Please let me know if it can be phrased better.

**Review by tupui — APPROVED** [review]
Thank you @aglebov for all your great work with this PR 👏 🚀 

This is all good to me and I am happy to get this in as it is. Tests are good, I check them locally as well, the documentation is also good with me. We can always tweak things later if we want.

We identified at least one follow up: using the new function to check for points in an hypercube. Let me know if you are interested in working on that.

For the order of the metric itself, as I said above we still have 5 months to change our minds. My initial polling is matching the current behaviour.

**Comment by tupui** [comment]
@aglebov for the protocol could you send a message to the mailing list? We have a silent agreement period of a week to observe.

**Comment by aglebov** [comment]
> @aglebov for the protocol could you send a message to the mailing list?

Done - awaiting moderator approval.

**Comment by mdhaber** [comment]
Thanks @aglebov. Could you summarize in the top post why this should be a new function rather than adding methods to `discrepancy`? That could be a common question.

**Comment by aglebov** [comment]
> Thanks @aglebov. Could you summarize in the top post why this should be a new function rather than adding methods to `discrepancy`? That could be a common question.

Done.

**Comment by mdhaber** [comment]
No disagreement from the mailing list, and this has the approval of an appropriate maintainer. Merge at will, @tupui.

**Review by tupui — APPROVED** [review]
All right, I am back and merging then 😃 Thank you again @aglebov and Matt for your help as well.

File: scipy/stats/tests/test_qmc.py
Comment (mdhaber) [inline scipy/stats/tests/test_qmc.py:232]: @tupui is this intended to `xfail` because of the warning (`UserWarning: Sample contains duplicate points.`)? If so, should it use `pytest.warns` to check for the warning (and pass)?

Comment (mdhaber) [reply scipy/stats/tests/test_qmc.py:232]: No, looks like it doesn't pass, but I am suppressing the warning.

Comment (tupui) [reply scipy/stats/tests/test_qmc.py:232]: IIRC to pass it needs a fix in the underlying voronoï construction.

Comment (aglebov) [reply scipy/stats/tests/test_qmc.py:232]: Looking at the `reason` field, the `xfail` is due to #18892 which remains unresolved.

**Pull request merged**
